File: Emory_MC/EnergyTransfer_Tm.py
import numpy as np
from Tm_inf import *
from Tm_adjustable_parameter import *


ET_n_term = ((n**2+2) / (3*n) )**4

# double the RME dictionary
for key in list(Tm_RME.keys()):
   parts = key.split('E')
   new_key = f'E{parts[2]}E{parts[1]}'
   if new_key not in Tm_RME:
      Tm_RME[new_key] = Tm_RME[key]



# degeneracy of energy level
Tm_g= {}
for key in Tm_energy.keys():

   key_symbol= key[key.find('(') + 1:-1] # 3H6
   J = int(key_symbol[-1]) / 2
   g = 2*J+1
   simplified_key = key[0 : key.find('(')]

   Tm_g[simplified_key] = g

# like this
# Tm_g={'E0':13, 'E1':9, 'E2':11, 'E3':9, 'E4':7, 'E5': 5, 'E6': 9, 'E7': 5, 'E8': 1, 'E9': 13, 'E10': 3, 'E11': 5}


Tm_energy_simplified = {}

for key, value in Tm_energy.items():
   simplified_key = key[0 : key.find('(')]
   Tm_energy_simplified[simplified_key] = value

# like this:
# Tm_energy = {'E0':153, 'E1': 5828, 'E2': 8396, 'E3':12735, 'E4':14598, 'E5':15180, 'E6': 21352, 'E7':28028, 'E8': 34900, 'E9': 35500, 'E10': 36400, 'E11': 38250}









class EnergyTransfer():
   def total_probability(self, r):
      pass

   def select_path(self, r):
      pass

   def add_state(self, ion12, ion22, rate):
      pass




class UpConversion(EnergyTransfer):
   def __init__ (self, ion2):
      self.ion2 = ion2
      self.resulting_states = []

   def total_probability(self, r): # TODO: 要不要在这里乘以10**（-6）
      return sum([result1[2]/(r/10**7)**6 for result1 in self.resulting_states])

   def select_path(self, r):
      if len(self.resulting_states) == 0 :
         return None
      
      results = [result1[2]/r**6 for result1 in self.resulting_states]
      results = [prob / sum(results) for prob in results]

      new_state = np.random.choice([i for i in range(len(self.resulting_states))], p=results)
      return self.resulting_states[new_state][0:2]
   
   def add_state(self, ion12, ion22, rate):
      self.resulting_states.append((ion12, ion22, rate))



def up_conversion():


   ion1_energy = 'S1'
   E_level = Tm_energy_simplified
   RME_value = Tm_RME
   g_value = Tm_g
   Omega_value = Tm_omega
   
   ret = {}
   for ion2_energy in Tm_energy_simplified:

      ion2_parts = ion2_energy.split('E')
      ion2_initial_state = int(ion2_parts[1])
      ion2_et = UpConversion(ion2_initial_state)

      all_transitions = {}
      delta_E_ion1 = {'S1S0': 10246}
      delta_E_ion2 = {f'{ion2_energy}{level}':  E_level[ion2_energy] -  E_level[level] for level in E_level if level != ion2_energy}
   
      for transition2, energy_diff2 in delta_E_ion2.items():

         ################################################################# important: ET matching condition
         if (energy_diff2 < 0 and abs(10246 + energy_diff2)< n_phonon*E_phonon):
            Delta_E = abs(10246 + energy_diff2)
            key = f'S1S0-{transition2}'
            all_transitions[key] = Delta_E
         
      for ET_key in all_transitions:
            first_part, second_part = ET_key.split('-')
            second_values = RME_value[second_part]
            all_transitions[ET_key] = [all_transitions[ET_key], 2*10**-20, second_values]
            parts = second_part.split('E')
            new_key = f'E{parts[1]}'
            all_transitions[ET_key].append([Yb_g[ion1_energy], g_value[new_key]])
      

      for key, value in all_transitions.items():

         S1 = value[1]
         S2 = Omega_value['2']*value[2][0]+Omega_value['4']*value[2][1]+Omega_value['6']*value[2][2]

         my_value = ET_constant*ET_n_term*s0*(S1*S2)*np.exp(-beta*value[0])/(value[3][0]*value[3][1])

         if (my_value > threshold):
            
            donor_transition, acceptor_transition = key.split('-')
            donor_parts = donor_transition.split('S')
            donor_final_state = int(donor_parts[2])
            acceptor_parts = acceptor_transition.split('E')
            acceptor_final_state = int(acceptor_parts[2])

            ion2_et.add_state(donor_final_state, acceptor_final_state, my_value)
            # ion2_et.add_state(0, int(key[-1]), my_value)

      

      ret[int(ion2_energy[1:])] = ion2_et

      
   return ret



class CrossRelaxation(EnergyTransfer):

   def __init__ (self, ion1, ion2):
      self.ion1 = ion1
      self.ion2 = ion2
      self.resulting_states = []

   def total_probability(self, r):
      return sum([result1[2]/(r/10**7)**6 for result1 in self.resulting_states])

   def select_path(self, r):
      if len(self.resulting_states) == 0 :
         return None
      
      results = [result1[2]/(r/10**7)**6 for result1 in self.resulting_states]
      results = [prob / sum(results) for prob in results]
      # print(self.resulting_states, total_prob, results)
      new_state = np.random.choice([i for i in range(len(self.resulting_states))], p=results)
      return self.resulting_states[new_state][0:2]
   
   def add_state(self, ion12, ion22, rate):
      self.resulting_states.append((ion12, ion22, rate))



def cross_relaxation():


   energy_levels = Tm_energy_simplified
   RME_value = Tm_RME
   g_value = Tm_g # deneracy
   Omega_value = Tm_omega

   ret = {}

   for ion1_energy in Tm_energy_simplified:
      ion1_ets = {}

      donor_parts = ion1_energy.split('E')
      donor_initial_state=int(donor_parts[1])


      for ion2_energy in Tm_energy_simplified:

         acceptor_parts = ion2_energy.split('E')
         acceptor_initial_state=int(acceptor_parts[1])

         ion1_ion2_et = CrossRelaxation(donor_initial_state, acceptor_initial_state)

         all_transitions = {}
         delta_E_ion1 = {f'{ion1_energy}{level}':  energy_levels[ion1_energy] -  energy_levels[level] for level in energy_levels if level != ion1_energy}
         delta_E_ion2 = {f'{ion2_energy}{level}': energy_levels[ion2_energy] - energy_levels[level] for level in energy_levels if level != ion2_energy}

         for transition1, energy_diff1 in delta_E_ion1.items():
            for transition2, energy_diff2 in delta_E_ion2.items():

               ################################################################# important: ET matching condition
               if (energy_diff1 > 0 and energy_diff2 < 0 and abs(energy_diff1+energy_diff2)< n_phonon*E_phonon) or (energy_diff1 < 0 and energy_diff2 > 0 and abs(energy_diff2+energy_diff1)< n_phonon*E_phonon):
                  
                  Delta_E = abs(energy_diff1 + energy_diff2)
                  key=f'{transition1}-{transition2}'
                  all_transitions[key]=Delta_E

         for ET_key in all_transitions:
            first_part, second_part = ET_key.split('-')
            
            # Oscillator strength: S (Emory)
            first_values = RME_value.get(first_part, [])
            second_values = RME_value.get(second_part, [])

            if first_values and second_values:

                  all_transitions[ET_key] = [all_transitions[ET_key], first_values, second_values]

            parts = first_part.split('E')
            new_key1 = f'E{parts[1]}'

            parts = second_part.split('E')
            new_key2 = f'E{parts[1]}'

            # append degeneracy of initial level of donor and acceptor
            all_transitions[ET_key].append([g_value[new_key1], g_value[new_key2]])

         for key, value in all_transitions.items():
            
            S1 = Omega_value['2']*value[1][0]+Omega_value['4']*value[1][1]+Omega_value['6']*value[1][2]
            S2 = Omega_value['2']*value[2][0]+Omega_value['4']*value[2][1]+Omega_value['6']*value[2][2]

            my_value = ET_constant*ET_n_term*s0*(S1*S2)*np.exp(-beta*value[0])/(value[3][0]*value[3][1])

            if (my_value > threshold):
               
               donor_transition, acceptor_transition = key.split('-')
               donor_parts = donor_transition.split('E')
               donor_final_state = int(donor_parts[2])
               acceptor_parts = acceptor_transition.split('E')
               acceptor_final_state = int(acceptor_parts[2])

               ion1_ion2_et.add_state(donor_final_state, acceptor_final_state, my_value)
               # ion1_ion2_et.add_state(int(key[3]), int(key[-1]), my_value)

         parts_ion2 = ion2_energy.split('E')
         ion2_initial_state = int(parts_ion2[1])
         ion1_ets[ion2_initial_state] = ion1_ion2_et
         # ion1_ets[int(ion2_energy[1])] = ion1_ion2_et


      parts_ion1 = ion1_energy.split('E')
      ion1_initial_state = int(parts_ion1[1])
      ret[ion1_initial_state] = ion1_ets
      #ret[int(ion1_energy[1])] = ion1_ets

      
   return ret

File: Emory_MC_cpp/EnergyTransfer_Tm.cpp
#include "utils.hpp"
#include "EnergyTransfer_Tm.hpp"
#include "Tm_inf.hpp"
#include "Tm_adjustable_parameter.hpp"
#include <cmath>
#include <unordered_map>
#include <vector>
#include <string>
#include <iostream>
#include <random>

static double ET_n_term;
static std::unordered_map<std::string, double> Tm_energy_simplified;
static std::unordered_map<std::string, int> Tm_g;
std::unordered_map<std::string, std::vector<double>> Tm_RME_tmp;

void initialize_data() {
    ET_n_term = std::pow(((TmAdjustableParameter::n * TmAdjustableParameter::n + 2) / (3.0 * TmAdjustableParameter::n)), 4);
    
    Tm_RME_tmp.clear();
    Tm_g.clear();
    Tm_energy_simplified.clear();
    Tm_RME_tmp = Tm_RME; 

    for (const auto& key_value : Tm_RME) {
        std::string key = key_value.first;
        auto firstE = key.find('E');
        auto secondE = key.find('E', firstE+1);
        std::string new_key = key.substr(secondE) + key.substr(firstE, secondE-firstE);
        if (Tm_RME.find(new_key) == Tm_RME.end()) {
            Tm_RME_tmp[new_key] = key_value.second;
        }
    }

    for (const auto& kv : Tm_energy) {
        std::string key = kv.first;
        auto pos = key.find('(');
        std::string simplified_key = key.substr(0, pos);
        double J = std::stod(key.substr(pos + 3, 1)) / 2;
        Tm_g[simplified_key] = 2 * J + 1;
        Tm_energy_simplified[simplified_key] = kv.second;
    }
}

UpConversion::UpConversion() : ion2(0) {} 

UpConversion::UpConversion(int ion2) : ion2(ion2) {}

UpConversion::UpConversion(const UpConversion& other) : ion2(other.ion2) {}

double UpConversion::total_probability(double r) const {
    double sum = 0.0;
    for (const auto& result : resulting_states) {
        sum += std::get<2>(result) / std::pow(r / 1e7, 6);
    }
    return sum;
}

std::pair<int, int> UpConversion::select_path(double r) {
    if (resulting_states.empty()) {
        return {-1, -1};
    }

    std::vector<double> results;
    for (const auto& result : resulting_states) {
        results.push_back(std::get<2>(result) / std::pow(r, 6));
    }
    
    double sum = std::accumulate(results.begin(), results.end(), 0.0);
    for (auto& prob : results) {
        prob /= sum;
    }

    std::random_device rd;
    std::mt19937 gen(rd());
    std::discrete_distribution<> d(results.begin(), results.end());
    int new_state_index = d(gen);

    return {std::get<0>(resulting_states[new_state_index]), std::get<1>(resulting_states[new_state_index])};
}

void UpConversion::add_state(int ion12, int ion22, double rate) {
    resulting_states.push_back({ion12, ion22, rate});
}


CrossRelaxation::CrossRelaxation() : ion1(0), ion2(0) {} 

CrossRelaxation::CrossRelaxation(int ion1, int ion2) : ion1(ion1), ion2(ion2) {}

CrossRelaxation::CrossRelaxation(const CrossRelaxation& other) : ion1(other.ion1), ion2(other.ion2) {}

double CrossRelaxation::total_probability(double r) const {
    double sum = 0.0;
    for (const auto& result : resulting_states) {
        sum += std::get<2>(result) / std::pow(r / 1e7, 6);
    }
    return sum;
}

std::pair<int, int> CrossRelaxation::select_path(double r) {
    if (resulting_states.empty()) return {-1, -1};

    std::vector<double> results;
    for (const auto& result : resulting_states) {
        results.push_back(std::get<2>(result) / std::pow(r, 6));
    }

    double sum = std::accumulate(results.begin(), results.end(), 0.0);
    for (auto& prob : results) {
        prob /= sum;
    }

    std::random_device rd;
    std::mt19937 gen(rd());
    std::discrete_distribution<> d(results.begin(), results.end());
    int new_state_index = d(gen);
    
    return {std::get<0>(resulting_states[new_state_index]), std::get<1>(resulting_states[new_state_index])};
}

void CrossRelaxation::add_state(int ion12, int ion22, double rate) {
    resulting_states.push_back({ion12, ion22, rate});
}


std::unordered_map<int, UpConversion> up_conversion() {
    std::unordered_map<int, UpConversion> ret;
    std::string ion1_energy = "S1";
    initialize_data();
    auto& E_level = Tm_energy_simplified;
    auto& RME_value = Tm_RME_tmp;
    auto& g_value = Tm_g;
    auto& Omega_value = Tm_omega;

    for (const auto& [ion2_energy, _] : Tm_energy_simplified) {
        auto pos = ion2_energy.find('E');
        int ion2_initial_state = std::stoi(ion2_energy.substr(pos + 1));
        UpConversion ion2_et(ion2_initial_state);
                
        for (const auto& [level, energy] : E_level) {
            if (level != ion2_energy) {
                double energy_diff2 = E_level[ion2_energy] - energy;
                if (energy_diff2 < 0 && std::abs(10246 + energy_diff2) < TmAdjustableParameter::n_phonon * TmAdjustableParameter::E_phonon) {
                    double Delta_E = std::abs(10246 + energy_diff2);
                    auto second_part = ion2_energy+level;
                    auto second_values = RME_value.at(ion2_energy+level);
                    auto new_key = level;

                    double S1 = 2e-20;
                    double S2 = Omega_value.at("2") * second_values[0] + Omega_value.at("4") * second_values[1] + Omega_value.at("6") * second_values[2];
                    double my_value = TmAdjustableParameter::ET_constant * ET_n_term * TmAdjustableParameter::s0 * (S1 * S2) * std::exp(-TmAdjustableParameter::beta * Delta_E) / (Yb_g.at(ion1_energy) * g_value.at(new_key));
                    if (my_value > TmAdjustableParameter::threshold) {
                        int donor_final_state = 0;
                        int acceptor_final_state = std::stoi(level.substr(level.find('E') + 1));

                        ion2_et.add_state(donor_final_state, acceptor_final_state, my_value);
                    }
                }
            }
        }
        
        ret[ion2_initial_state] = ion2_et;
    }

    return ret;
}


std::unordered_map<int, std::unordered_map<int, CrossRelaxation>> cross_relaxation() {
    std::unordered_map<int, std::unordered_map<int, CrossRelaxation>> ret;
    initialize_data();
    auto& E_level = Tm_energy_simplified;
    auto& RME_value = Tm_RME_tmp;
    auto& g_value = Tm_g;
    auto& Omega_value = Tm_omega;
    
    for (const auto& ion1_energy : Tm_energy_simplified) {
        std::unordered_map<int, CrossRelaxation> ion1_ets;
        auto pos = ion1_energy.first.find('E');
        int donor_initial_state = std::stoi(ion1_energy.first.substr(pos + 1));
        
        for (const auto& ion2_energy : Tm_energy_simplified) {

            pos = ion2_energy.first.find('E');
            int acceptor_initial_state = std::stoi(ion2_energy.first.substr(pos + 1));

            CrossRelaxation ion1_ion2_et(donor_initial_state, acceptor_initial_state);

            std::unordered_map<std::string, double> all_transitions;
            for (const auto& level : Tm_energy_simplified) {
                if (level.first != ion1_energy.first) {
                    double energy_diff1 = Tm_energy_simplified.at(ion1_energy.first) - level.second;
                    std::string transition1 = ion1_energy.first + level.first;
                    
                    for (const auto& level2 : Tm_energy_simplified) {
                        if (level2.first != ion2_energy.first) {
                            double energy_diff2 = Tm_energy_simplified.at(ion2_energy.first) - level2.second;
                            std::string transition2 = ion2_energy.first + level2.first;

                            if ((energy_diff1 > 0 && energy_diff2 < 0 && std::abs(energy_diff1 + energy_diff2) < TmAdjustableParameter::n_phonon * TmAdjustableParameter::E_phonon) ||
                                (energy_diff1 < 0 && energy_diff2 > 0 && std::abs(energy_diff2 + energy_diff1) < TmAdjustableParameter::n_phonon * TmAdjustableParameter::E_phonon)) {
                                all_transitions[transition1 + "-" + transition2] = std::abs(energy_diff1 + energy_diff2);

                                const auto& first_values = RME_value.at(transition1);
                                const auto& second_values = RME_value.at(transition2);

                                if (!first_values.empty() && !second_values.empty()) {
                                    double S1 = Tm_omega.at("2") * first_values[0] + Tm_omega.at("4") * first_values[1] + Tm_omega.at("6") * first_values[2];
                                    double S2 = Tm_omega.at("2") * second_values[0] + Tm_omega.at("4") * second_values[1] + Tm_omega.at("6") * second_values[2];
                                    double my_value = TmAdjustableParameter::ET_constant * ET_n_term * TmAdjustableParameter::s0 * (S1 * S2) 
                                                      * std::exp(-TmAdjustableParameter::beta * abs(energy_diff1 + energy_diff2)) / (g_value.at(level.first) * g_value.at(level2.first));

                                    if (my_value > TmAdjustableParameter::threshold){
                                        int donor_final_state = std::stoi(transition1.substr(transition1.find('E') + 1));
                                        int acceptor_final_state = std::stoi(transition2.substr(transition2.find('E') + 1));
                                        ion1_ion2_et.add_state(donor_final_state, acceptor_final_state, my_value);
                                    }
                                }
                            }
                        }
                    }
                }
            }

            ion1_ets[acceptor_initial_state] = ion1_ion2_et;
        }

        ret[donor_initial_state] = ion1_ets;
    }

    return ret;
}